### PR TITLE
Adds is_fresh predicate for function contracts

### PR DESCRIPTION
**Pull request by feliperodri**
Signed-off-by: Felipe R. Monteiro <felisous@amazon.com>

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

**Comment by martin-cs** [comment]
( At some point before the "don't review" tag gets removed, please can this PR say what the semantics of `is_fresh` is. )

File: src/ansi-c/library/cprover.h [hunk 1]
```diff
@@ -53,6 +53,9 @@ void __CPROVER_printf(const char *format, ...);
 void __CPROVER_input(const char *id, ...);
 void __CPROVER_output(const char *id, ...);
 
+// contract-related functions
+__CPROVER_bool __CPROVER_is_fresh(const void *mem, __CPROVER_size_t size);
+
```
Comment (SaswatPadhi) [inline src/ansi-c/library/cprover.h]: Nice. Can we also move `__CPROVER_old` from line 39 to this section?

File: regression/contracts/quantifiers-loop-02/test.desc [hunk 2]
```diff
@@ -3,9 +3,9 @@ main.c
 --enforce-all-contracts _ --z3
 ^EXIT=0$
 ^SIGNAL=0$
-^\[main.1\] line .* Check loop invariant before entry: SUCCESS
-^\[main.2\] line .* Check that loop invariant is preserved: SUCCESS
-^\[main.assertion.1\] line .* assertion .*: SUCCESS
+^\[main.\d+\] line .* Check loop invariant before entry: SUCCESS
+^\[main.\d+\] line .* Check that loop invariant is preserved: SUCCESS
+^\[main.assertion.\d+\] line .* assertion .*: SUCCESS
```
Comment (SaswatPadhi) [inline regression/contracts/quantifiers-loop-02/test.desc:8]: Thanks for getting rid of these hard-coded numbers.
Could please split off this (and any other cleanup changes that you make) to a separate commit?

I also think I should tweak the comment for the loop invariant assertions at some point. From `[main.1` to `[main.invariant.1]` :)

File: regression/contracts/is_unique_01_replace/main.c [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+{
+  int a;
+  a = *x;
+  return ret_value == *x + 5;
+}
+
+// The 'ensures' __CPROVER_is_fresh test is unnecessary, but left in just to test the function is working correctly.
+// If you remove the negation, the program will fail, because 'x' is not fresh.
+
+int foo(int *x, int y) __CPROVER_assigns(*x)
```
Comment (SaswatPadhi) [inline regression/contracts/is_unique_01_replace/main.c:65]: Please use clang-format on/off around these contract annotations. We have used them in other test files and I think it does make the contracts more readable. Please see `quantifiers-exists-ensures-enforce` for example.

File: src/ansi-c/cprover_builtin_headers.h [hunk 4]
```diff
@@ -35,6 +35,9 @@ void __CPROVER_atomic_begin();
 void __CPROVER_atomic_end();
 void __CPROVER_fence(const char *kind, ...);
 
+// contract-related functions
+__CPROVER_bool __CPROVER_is_fresh(const void *mem, __CPROVER_size_t size);
+
```
Comment (SaswatPadhi) [inline src/ansi-c/cprover_builtin_headers.h:40]: This looks like a duplicate of the change in `src/ansi-c/library/cprover.h`. Do we need to have this in both files?

File: regression/contracts/test_scalar_memory_replace/test.desc [hunk 5]
```diff
@@ -0,0 +1,13 @@
...
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that verification succeeds if only expressions inside the assigns clause are assigned within the function.
+
+Note: For all 'enforce' tests, nothing can be assumed about the return value of the function (as the function call is not replaced at this point).
```
Comment (SaswatPadhi) [inline regression/contracts/test_scalar_memory_replace/test.desc]: I don't understand this sentence: for 'enforce' tests, the function call is not replaced with its contract, but should still behave as expected when called. And that's the reason why `assert(o >= 10 && o == *n + 5);` works in your code.

So may be just remove this line (and the next line)? Or, try to clarify the message here.

File: regression/contracts/test_scalar_memory_replace/test.desc [hunk 6]
```diff
@@ -0,0 +1,13 @@
+CORE
+main.c
+--replace-all-calls-with-contracts
+^EXIT=0$
+^SIGNAL=0$
+^VERIFICATION SUCCESSFUL$
+--
+--
+This test checks that verification succeeds if only expressions inside the assigns clause are assigned within the function.
```
Comment (SaswatPadhi) [inline regression/contracts/test_scalar_memory_replace/test.desc]: I think this is an old message. Please take a pass over all the `test.desc` files and see that the message corresponds to the behaviro we're testing for.

File: regression/contracts/test_scalar_memory_replace/main.c [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+
+int main()
+{
+  int *n = malloc(sizeof(int));
+  assert(__CPROVER_r_ok(n, sizeof(int)));
+  *n = 3;
+  int o = foo(n);
+  assert(o >= 10 && o == *n + 5);
+  return 0;
+}
```
Comment (SaswatPadhi) [inline regression/contracts/test_scalar_memory_replace/main.c]: Missing newline at the end. Please also check other files.

File: regression/contracts/test_scalar_memory_replace/main.c [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+// If you remove the negation, the program will fail, because 'x' is not fresh.
+
+int foo(int *x) __CPROVER_assigns(*x)
+  __CPROVER_requires(__CPROVER_is_fresh(x, sizeof(int)) && *x > 0 && ptr_ok(x))
+    __CPROVER_ensures(
+      !ptr_ok(x) && !__CPROVER_is_fresh(x, sizeof(int)) &&
+      return_ok(__CPROVER_return_value, x))
+{
+  *x = *x + 4;
+  int y = *x + 5;
```
Comment (SaswatPadhi) [inline regression/contracts/test_scalar_memory_replace/main.c]: May be remove this line? Please also remove unused headers at the top.

Let's keep the each regression test small and very specific to the target behavior being tested.

File: regression/contracts/test_multifile_replace/main.c [hunk 9]
```diff
@@ -0,0 +1,20 @@
+#include "foo.h"
```
Comment (SaswatPadhi) [inline regression/contracts/test_multifile_replace/main.c]: `#include` get replaced in the same source file during preprocessing, so I am not sure what is the objective of splitting them.

It would make sense to test contracts within multiple translation units, but we are not doing that here. Same for all the other header files.

**Review by SaswatPadhi — COMMENTED** [review]
I took a first pass over the regression tests. I left some small comments below, but here are some major ones:

- The regression tests all look kind of similar. Some other corner cases that we should test (feel free to add more):
  - requires(is_fresh()) should pass on whole structs and individual members
  - ensures(is_fresh()) should pass on functions that perform allocation directly and return a new object
  - ensures(is_fresh()) should pass on functions that call another function that does the allocation and then return the object (this is to check that we propagate freshness correctly across contracts)

I will review the implementation logic in the next pass, and finally the code style in a later pass.

One high-level issue that I am confused about:

```c
// When enforced, this contract should pass
bool sub_ptr_values(int *x, int *y)
  __CPROVER_requires(
    __CPROVER_is_fresh(x, sizeof(int)) && __CPROVER_is_fresh(y, sizeof(int))
  )
  __CPROVER_ensures(
    __CPROVER_return_value == *x - *y
  )
{
  return *x - *y;
}

// A function that uses `sub_ptr_values`
void foo()
{
  int *n = malloc(sizeof(int));

  // When call is replaced by contract, this will cause an assertion violation?
  int diff = sub_ptr_values(n, n);
}
```

The question here is, how should the contract `sub_ptr_values` be written so as to allow _possibly-aliased_ arguments? I see "is_fresh" as enforcing/checking fresh non-aliasing allocations.

Okay I guess one could make the aliasing explicit:
```c
  __CPROVER_requires(
    __CPROVER_is_fresh(x, sizeof(int)) &&
    (y == x || __CPROVER_is_fresh(y, sizeof(int)))
  )
```
Can we add this conditional "is_fresh" case as a regression test?

File: src/ansi-c/library/cprover.h (hunk 1 above)
Comment (feliperodri) [reply src/ansi-c/library/cprover.h]: We don't need these signatures at all here.

File: regression/contracts/quantifiers-loop-02/test.desc (hunk 2 above)
Comment (feliperodri) [reply regression/contracts/quantifiers-loop-02/test.desc:8]: I added the clean-ups on a separate PR. +1 to the markup for invariants!

File: regression/contracts/is_unique_01_replace/main.c (hunk 3 above)
Comment (feliperodri) [reply regression/contracts/is_unique_01_replace/main.c:65]: I added clang-format on/off to all regression tests.

File: src/ansi-c/cprover_builtin_headers.h (hunk 4 above)
Comment (feliperodri) [reply src/ansi-c/cprover_builtin_headers.h:40]: We don't need to add anything to `src/ansi-c/library/cprover.h`. I've group the contract-related function declarations here.

File: regression/contracts/test_scalar_memory_replace/test.desc (hunk 5 above)
Comment (feliperodri) [reply regression/contracts/test_scalar_memory_replace/test.desc]: This is a old message, I re-wrote it.

File: regression/contracts/test_scalar_memory_replace/test.desc (hunk 6 above)
Comment (feliperodri) [reply regression/contracts/test_scalar_memory_replace/test.desc]: I took a pass over the test descriptions to re-write all the messages.

File: regression/contracts/test_scalar_memory_replace/main.c (hunk 7 above)
Comment (feliperodri) [reply regression/contracts/test_scalar_memory_replace/main.c]: Fixed it.

File: regression/contracts/test_scalar_memory_replace/main.c (hunk 8 above)
Comment (feliperodri) [reply regression/contracts/test_scalar_memory_replace/main.c]: Removed it. I also cleaned up includes.

File: regression/contracts/test_multifile_replace/main.c (hunk 9 above)
Comment (feliperodri) [reply regression/contracts/test_multifile_replace/main.c]: Before, we added function contracts on function declarations and place them on `.h` files, which would make sense to test here. This is not the case anymore. We can test/add this feature on a separate PR. I remove all regression tests design for multiple files (from this PR).

**Comment by feliperodri** [comment]
@SaswatPadhi I addressed all your comments and added more regression tests. Could you take another look at it?

File: regression/contracts/quantifiers-loop-02/test.desc (hunk 2 above)
Comment (SaswatPadhi) [reply regression/contracts/quantifiers-loop-02/test.desc:8]: Thank you. Resolving this.

File: regression/contracts/test_scalar_memory_replace/test.desc (hunk 6 above)
Comment (SaswatPadhi) [reply regression/contracts/test_scalar_memory_replace/test.desc]: Thank you.

File: src/ansi-c/library/contracts.c [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+}
+
+/* FUNCTION: __CPROVER_FUN_NAME_call_ensures_is_fresh */
+_Bool __CPROVER_FUN_NAME_call_ensures_is_fresh(
+  void **elem,
+  __CPROVER_size_t size)
+{
+  *elem = malloc(size);
+  return (*elem != NULL);
+}
```
Comment (SaswatPadhi) [inline src/ansi-c/library/contracts.c]: [minor]

Missing newline.

File: regression/contracts/test_struct_replace/test.desc [hunk 11]
```diff
@@ -0,0 +1,13 @@
...
+^EXIT=0$
+^SIGNAL=0$
+\[precondition.\d+\] file main.c line \d+ Check requires clause: SUCCESS
+\[main.assertion.\d+\] line \d+ assertion rval \=\= x-\>baz \+ x-\>qux: SUCCESS
+\[main.assertion.\d+\] line \d+ assertion \*x \=\= \*y: SUCCESS
+^VERIFICATION SUCCESSFUL$
+--
+--
+Checks whether asserting __CPROVER_is_fresh will guarantee that objects
+mapped to distinct objetcs (for structs).
```
Comment (SaswatPadhi) [inline regression/contracts/test_struct_replace/test.desc]: [minor]

Missing newline.

File: regression/contracts/test_struct_member_enforce/test.desc [hunk 12]
```diff
@@ -0,0 +1,13 @@
...
+^EXIT=0$
+^SIGNAL=0$
+\[postcondition.\d+\] file main.c line \d+ Check ensures clause: SUCCESS
+\[foo.\d+\] line \d+ Check that x-\>str\[\(.*\)\(x-\>len - 1\)\] is assignable: SUCCESS
+\[main.assertion.\d+\] line \d+ assertion rval \=\= 128: SUCCESS
+^VERIFICATION SUCCESSFUL$
+--
+--
+Checks whether assuming __CPROVER_is_fresh will guarantee a new freshly
+allocated pointer (no aliasing) for struct members.
```
Comment (SaswatPadhi) [inline regression/contracts/test_struct_member_enforce/test.desc]: [minor]

Missing newline.

File: regression/contracts/test_struct_enforce/test.desc [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+^SIGNAL=0$
+\[postcondition.\d+\] file main.c line \d+ Check ensures clause: SUCCESS
+\[foo.\d+\] line \d+ Check that x-\>baz is assignable: SUCCESS
+\[foo.\d+\] line \d+ Check that x-\>qux is assignable: SUCCESS
+\[main.assertion.\d+\] line \d+ assertion rval \=\= 10: SUCCESS
+^VERIFICATION SUCCESSFUL$
+--
+--
+Checks whether assuming __CPROVER_is_fresh will guarantee a new freshly
+allocated pointer (no aliasing) for structs.
```
Comment (SaswatPadhi) [inline regression/contracts/test_struct_enforce/test.desc]: [minor]

Missing newline.

File: regression/contracts/test_scalar_memory_replace/main.c (hunk 7 above)
Comment (SaswatPadhi) [reply regression/contracts/test_scalar_memory_replace/main.c]: Sorry, but it looks like this file is still missing a newline at the end?
I did refresh the PR page and I am looking at the current changes.

File: regression/contracts/test_array_memory_too_small_replace/main.c [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+}
+
+int main()
+{
+  int *n = malloc(sizeof(int) * 9);
+  n[0] = 3;
+  int o = foo(n);
+  assert(o >= 10 && o == *n + 5);
+  return 0;
+}
```
Comment (SaswatPadhi) [inline regression/contracts/test_array_memory_too_small_replace/main.c]: [minor]

Missing newline.

File: regression/contracts/test_array_memory_replace/main.c [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+int main()
+{
+  int *n = malloc(sizeof(int) * 10);
+  n[0] = 3;
+  int o = foo(n);
+  assert(o >= 10 && o == *n + 5);
+  assert(n[9] == 113);
+  return 0;
+}
```
Comment (SaswatPadhi) [inline regression/contracts/test_array_memory_replace/main.c]: [minor]

Missing newline.

File: regression/contracts/test_aliasing_replace/test.desc [hunk 16]
```diff
@@ -0,0 +1,12 @@
...
+--replace-all-calls-with-contracts
+^EXIT=10$
+^SIGNAL=0$
+\[precondition.\d+\] file main.c line \d+ Check requires clause: FAILURE
+\[main.assertion.\d+\] line \d+ assertion \!\(n \< 4\): SUCCESS
+^VERIFICATION FAILED$
+--
+--
+Checks whether asserting __CPROVER_is_fresh will guarantee that objects
+mapped to distinct objetcs.
```
Comment (SaswatPadhi) [inline regression/contracts/test_aliasing_replace/test.desc]: [minor]

Missing newline.

File: src/ansi-c/library/contracts.c [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+_Bool __CPROVER_is_fresh(const void *ptr, __CPROVER_size_t size)
+{
+  __CPROVER_assert(
+    0, "__CPROVER_is_fresh called outside of contract function.");
+  return 0;
+}
+
+//
+// Below are templates of the functions that
+// will be instantiated for each contract function.
```
Comment (SaswatPadhi) [inline src/ansi-c/library/contracts.c]: Are these templates used?

I might be wrong, but it looks like we generate concrete instances of these templates during instrumentation, but from scratch.

File: src/goto-instrument/code_contracts.cpp [hunk 18]
```diff
@@ -14,53 +14,107 @@ Date: February 2016
...
-/// Predicate to be used with the exprt::visit() function. The function
-/// found_return_value() will return `true` iff this predicate is called on an
-/// expr that contains `__CPROVER_return_value`.
-class return_value_visitort : public const_expr_visitort
+#include <ansi-c/ansi_c_language.h>
+
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
+
+#include "loop_utils.h"
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Could you please evaluate all newly added `#include`s? I suspect this wasn't rebased against the recent version in which I removed some of these. Pretty sure `loop_utils.h` was one of them.

File: src/goto-instrument/code_contracts.cpp [hunk 19]
```diff
@@ -14,53 +14,107 @@ Date: February 2016
...
 #include <util/replace_symbol.h>
 
-/// Predicate to be used with the exprt::visit() function. The function
-/// found_return_value() will return `true` iff this predicate is called on an
-/// expr that contains `__CPROVER_return_value`.
-class return_value_visitort : public const_expr_visitort
+#include <ansi-c/ansi_c_language.h>
+
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Please place these 2 lines above, together with lines 25 -- 27, and then sort all of the `goto-programs/` headers.

File: src/goto-instrument/code_contracts.cpp [hunk 20]
```diff
@@ -14,53 +14,107 @@ Date: February 2016
...
 #include <util/pointer_offset_size.h>
 #include <util/pointer_predicates.h>
+#include <util/prefix.h>
 #include <util/replace_symbol.h>
 
-/// Predicate to be used with the exprt::visit() function. The function
-/// found_return_value() will return `true` iff this predicate is called on an
-/// expr that contains `__CPROVER_return_value`.
-class return_value_visitort : public const_expr_visitort
+#include <ansi-c/ansi_c_language.h>
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: Please place this line above, together with line 23.

File: src/goto-instrument/code_contracts.cpp [hunk 21]
```diff
@@ -14,53 +14,107 @@ Date: February 2016
...
+
+#include <goto-programs/goto_convert.h>
+#include <goto-programs/goto_model.h>
+
+#include "loop_utils.h"
+
+// TEMPORARY
```
Comment (SaswatPadhi) [inline src/goto-instrument/code_contracts.cpp]: What does temporary mean here?

**Review by SaswatPadhi — CHANGES_REQUESTED** [review]
Thanks for improving the regression tests! They were quite helpful for me to understand end-to-end behavior before reviewing the core logic.

I left a few more minor comments below. The contract implementation looks sounds to me. The only remaining major question I have is regarding data structures like lists. I am not sure how one would write a contract using `is_fresh` for a function that takes a linked list and say counts its length. I can see how that could be written for lists that have a constant length, but it's not obvious how to write it for bounded lists, not unbounded, but that do not have a constant length.

If inductive structures like lists are beyond the scope of this first version of code contracts, let me know.

**Comment by feliperodri** [comment]
@SaswatPadhi 

> I left a few more minor comments below. The contract implementation looks sounds to me. The only remaining major question I have is regarding data structures like lists. I am not sure how one would write a contract using `is_fresh` for a function that takes a linked list and say counts its length. I can see how that could be written for lists that have a constant length, but it's not obvious how to write it for bounded lists, not unbounded, but that do not have a constant length.

For inductive data structures, one needs to invoke `__CPROVER_is_fresh` for every node, which is not feasible for unbounded (and bounded perhaps) approaches. We could add more predicates for specific cases, e.g., `__CPROVER_list_is_fresh(list)` would propagate the property to every node of the list. This is difficult, because we would need to make assumption on the shape of the list.

> If inductive structures like lists are beyond the scope of this first version of code contracts, let me know.

This is out of the scope for this PR; however, we should talk more about it offline to discuss ideas on how we can address this in the future.

File: src/ansi-c/library/contracts.c (hunk 10 above)
Comment (feliperodri) [reply src/ansi-c/library/contracts.c]: Fixed it.

File: regression/contracts/test_struct_replace/test.desc (hunk 11 above)
Comment (feliperodri) [reply regression/contracts/test_struct_replace/test.desc]: Fixed it.

File: regression/contracts/test_struct_member_enforce/test.desc (hunk 12 above)
Comment (feliperodri) [reply regression/contracts/test_struct_member_enforce/test.desc]: Fixed it.

File: regression/contracts/test_struct_enforce/test.desc (hunk 13 above)
Comment (feliperodri) [reply regression/contracts/test_struct_enforce/test.desc]: Fixed it.

File: regression/contracts/test_array_memory_too_small_replace/main.c (hunk 14 above)
Comment (feliperodri) [reply regression/contracts/test_array_memory_too_small_replace/main.c]: Fixed it.

File: regression/contracts/test_array_memory_replace/main.c (hunk 15 above)
Comment (feliperodri) [reply regression/contracts/test_array_memory_replace/main.c]: Fixed it.

File: regression/contracts/test_aliasing_replace/test.desc (hunk 16 above)
Comment (feliperodri) [reply regression/contracts/test_aliasing_replace/test.desc]: Fixed it.

File: src/ansi-c/library/contracts.c (hunk 17 above)
Comment (feliperodri) [reply src/ansi-c/library/contracts.c]: We're not using these templates at the moment, but generating concrete instances during instrumentation as you observed. I removed this for now, so we can make use of these templates on a separate PR (as discussed offline).

File: src/goto-instrument/code_contracts.cpp (hunk 21 above)
Comment (feliperodri) [reply src/goto-instrument/code_contracts.cpp]: Unnecessary includes. I removed them.

File: src/goto-instrument/code_contracts.cpp (hunk 18 above)
Comment (feliperodri) [reply src/goto-instrument/code_contracts.cpp]: I reviewed all includes. This is garbage accumulated from the rebase process.

File: src/goto-instrument/code_contracts.cpp (hunk 19 above)
Comment (feliperodri) [reply src/goto-instrument/code_contracts.cpp]: Fixed it.

File: src/goto-instrument/code_contracts.cpp (hunk 20 above)
Comment (feliperodri) [reply src/goto-instrument/code_contracts.cpp]: Fixed it.

**Review by SaswatPadhi — APPROVED** [review]
Thanks! LGTM 😃

There are a bunch of TODOs but let's merge this first and then tackle the refactoring, documentation etc. in future PRs.

File: src/goto-instrument/code_contracts.cpp [hunk 22]
```diff
@@ -1809,3 +1875,277 @@ exprt assigns_clauset::compatible_expression(
...
+  ensures_visitor(ensures);
+  for(auto it : ensures_visitor.is_fresh_calls())
+  {
+    create_ensures_fn_call(it);
+  }
+}
+
+//
+//
+// Code largely copied from model_argc_argv.cpp
```
Comment (chrisr-diffblue) [inline src/goto-instrument/code_contracts.cpp:1933]: Is there any way to factor out the common code, rather than copying?

File: src/goto-instrument/code_contracts.cpp [hunk 23]
```diff
@@ -1809,3 +1875,277 @@ exprt assigns_clauset::compatible_expression(
...
+}
+
+is_fresh_replacet::is_fresh_replacet(
+  code_contractst &_parent,
+  messaget _log,
+  irep_idt _fun_id)
+  : is_fresh_baset(_parent, _log, _fun_id)
+{
+  std::stringstream ssreq, ssensure, ssmemmap;
+  ssreq /* << CPROVER_PREFIX */ << fun_id << "_call_requires_is_fresh";
```
Comment (chrisr-diffblue) [inline src/goto-instrument/code_contracts.cpp:2104]: Are these commented out CPROVER_PREFIX's deliberate? is_fresh_enforcet::is_fresh_enforcet seems to insert the, but they are commented out here - just wondering which is correct, or whether both are?

**Review by chrisr-diffblue — APPROVED** [review]
Nothing blocking from my point of view - though like @martin-cs it would be nice to see a clear description of the intended semantics - plus a couple of small queries/comments.
